File: payroll-api/src/main/java/com/company/payroll/filter/JwtAuthenticationFilter.java
package com.company.payroll.filter;

import java.io.IOException;
import java.util.ArrayList;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import com.company.payroll.util.JwtTokenUtils;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

/**
 * 
 * Created on 26 Apr 2023
 * 
 * <p> Updated at 1 May 2023. Include Spring default authorization mechanism for authenticate request made from users.
 */
@Component
public class JwtAuthenticationFilter extends OncePerRequestFilter {

	@Autowired
    private JwtTokenUtils jwtTokenUtils;
    
	@Override
	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
			throws ServletException, IOException {
		String header = request.getHeader("Authorization");
		
		// Token empty/null, pass the filter
		if(header == null || !header.startsWith("Bearer")) {
			filterChain.doFilter(request, response);
			return;
		}
		
		final String token = header.substring(7);
		
		// Token not valid, pass the filter
		if(!jwtTokenUtils.validateToken(token)) {
			filterChain.doFilter(request, response);
		}
		
		// Token exist and not expired, get username
		String username = jwtTokenUtils.getUsername(token);

		// Initialize authentication token
		UsernamePasswordAuthenticationToken authToken = new UsernamePasswordAuthenticationToken(username, null, new ArrayList<>());
		authToken.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));
		
		// Pass authentication token to security context
		SecurityContextHolder.getContext().setAuthentication(authToken);
		
		filterChain.doFilter(request, response);
	}
}

File: kotlin-payroll/src/main/kotlin/com/company/payroll/filter/JwtAuthenticationFilter.kt
package com.company.payroll.filter

import com.company.payroll.util.JwtTokenUtils
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
class JwtAuthenticationFilter(@Autowired private val jwtTokenUtils: JwtTokenUtils): OncePerRequestFilter() {
  override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
    // Token empty/null, pass the filter
    val header = request.getHeader("Authorization")
    if(header == null || header.isEmpty() || !header.startsWith("Bearer")) {
      filterChain.doFilter(request, response)
      return;
    }

    val token = header.substring(7)

    // Token not valid, pass the filter
    if(!jwtTokenUtils.validateToken(token)) {
      filterChain.doFilter(request, response)
    }

    // Token exist and not expired, get username
    val username = jwtTokenUtils.getUsername(token)

    // Initialize authentication token
    val authToken = UsernamePasswordAuthenticationToken(username, null, ArrayList())
    authToken.details = WebAuthenticationDetailsSource().buildDetails(request)

    // Pass authentication token to security context
    SecurityContextHolder.getContext().authentication = authToken

    filterChain.doFilter(request, response)
  }

}